File: 올바른 괄호.py
from collections import deque
def solution(s):
    answer = True
    d = deque()
    d.append(s[0])
    for i in range(1, len(s)):
        if d: top = d[-1]
        else: 
            d.append(s[i])
            continue
        if top == "(" and s[i] == ")": d.pop()
        else:
            d.append(s[i])
    if d: answer = False
    else: answer = True
    return answer

File: 올바른 괄호.java
import java.util.Deque;
import java.util.ArrayDeque;
class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Deque<Character> d = new ArrayDeque<>();
        
        for(int i = 0; i<s.length(); i++) {
            
            if (d.size() == 0){
                d.add(s.charAt(i));
                continue;
            }
            char top = d.peekLast();
            if(top == s.charAt(i)){
                d.add(s.charAt(i));
                continue;
            }else if(top == '(' && s.charAt(i) == ')'){
                d.pop();
                continue;
            }
        }
        if(d.size()>0) answer = false;

        return answer;
    }
}
